cmd/clusterawsadm/iamservice/service.go

```diff
@@ -0,0 +1,500 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package iamservice is a redefined way to create and delete IAM instance profiles, policies and roles.
+package iamservice
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"regexp"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/configservice"
+	"github.com/aws/aws-sdk-go/service/iam"
+	go_cfn "github.com/awslabs/goformation/v4/cloudformation"
+	cfn_iam "github.com/awslabs/goformation/v4/cloudformation/iam"
+	"github.com/pkg/errors"
+	"k8s.io/klog/v2"
+
+	iamv1 "sigs.k8s.io/cluster-api-provider-aws/v2/iam/api/v1beta1"
+)
+
+// Service defines methods for managing the AWS IAM resources using the AWS SDK.
+type Service interface {
+	CreateResources(t go_cfn.Template, tags map[string]string) error
+	DeleteResources(t go_cfn.Template, tags map[string]string) error
+}
+
+type serviceImpl struct {
+	IAM *iam.IAM
+}
+
+// New creates a new IAM service object to interact with the AWS SDK.
+func New(iamSvc *iam.IAM) Service {
+	return &serviceImpl{
+		IAM: iamSvc,
+	}
+}
+
+func createClient() (*iam.IAM, error) {
+	s, err := session.NewSession()
+	if err != nil {
+		return nil, errors.Wrap(err, "internal server error")
+	}
+	return iam.New(s), nil
+}
+
+func prioritySet(t go_cfn.Template) (rmap map[string][]go_cfn.Resource, err error) {
+	rmap = map[string][]go_cfn.Resource{}
+	for _, resource := range t.Resources {
+		if resource.AWSCloudFormationType() == configservice.ResourceTypeAwsIamRole {
```

Comment:
The description says that this PR is looking to move away from CloudFormation, but we're using CloudFormation files here.

Is the intent to use CloudFormation templates as the input still, but introspect them and make our own API calls vs submitting them directly to the CF service endpoints?

Reply:
The intent as discussed with my mentors Richard and Ankita was to think of a solution that eventually gets rid of the CloudFormation to manage IAM resources. The end goal is to create an idempotent solution that enforces API calls to AWS to manage the IAM resources and doesn't rely on CloudFormation Stack. And the intent is to do this incrementally and not all at once so that we avoid big code change altogether and maintain a low to no user side changes.

So this PR is the first step in achieving the above, where we rely on the CF config(template) to get the input and make API calls to manage these IAM resources without relying on CloudFormation stack.